File: frontend/src/lib/api/imaging.ts
/** Imaging API types and functions */
import { apiClient } from './client';

export interface ImagePreviewPayload {
	image: number[][];
	stats: {
		shape: number[];
		integrated_shape: number[];
		min: number;
		max: number;
		mean: number;
		std: number;
		cube_name: string;
	};
	method: string;
}

export interface DeconvolutionResponse {
	dirty: ImagePreviewPayload;
	component_model: ImagePreviewPayload;
	restored: ImagePreviewPayload;
	residual: ImagePreviewPayload;
	reference_clean: ImagePreviewPayload | null;
	convolved_reference: ImagePreviewPayload | null;
	metadata: {
		dirty_cube_path: string;
		beam_cube_path: string;
		beam_cube_name: string;
		clean_cube_path: string | null;
		cycles_requested: number;
		cycles_completed: number;
		cycles_added: number;
		gain: number;
		threshold: number | null;
		state_path: string;
		component_cube_path: string;
		restored_cube_path: string;
		residual_cube_path: string;
		convolved_reference_path: string | null;
		resumed: boolean;
	};
}

export const imagingApi = {
	deconvolve(params: {
		directory: string;
		dirtyCubePath: string;
		beamCubePath: string;
		cleanCubePath?: string | null;
		cycles?: number;
		gain?: number;
		threshold?: number | null;
		statePath?: string | null;
		method?: 'sum' | 'mean';
	}): Promise<DeconvolutionResponse> {
		return apiClient.post<DeconvolutionResponse>('/api/v1/imaging/deconvolve', {
			directory: params.directory,
			dirty_cube_path: params.dirtyCubePath,
			beam_cube_path: params.beamCubePath,
			clean_cube_path: params.cleanCubePath ?? null,
			cycles: params.cycles ?? 100,
			gain: params.gain ?? 0.1,
			threshold: params.threshold ?? null,
			state_path: params.statePath ?? null,
			method: params.method ?? 'sum'
		});
	}
};

File: backend/app/schemas/imaging.py
"""Schemas for imaging/deconvolution API endpoints."""

from typing import Any, Literal

from pydantic import BaseModel, Field


class DeconvolutionRequest(BaseModel):
    """Request body for iterative deconvolution from saved cube products."""

    directory: str = Field(..., description="Base directory for relative file paths")
    dirty_cube_path: str = Field(..., description="Relative or absolute path to the dirty cube")
    beam_cube_path: str = Field(..., description="Relative or absolute path to the beam cube")
    clean_cube_path: str | None = Field(
        default=None,
        description="Optional reference clean/model cube for visual comparison",
    )
    cycles: int = Field(default=100, ge=0, le=5000, description="Number of CLEAN cycles")
    gain: float = Field(default=0.1, gt=0.0, le=1.0, description="Loop gain")
    threshold: float | None = Field(
        default=None,
        ge=0.0,
        description="Optional absolute stopping threshold on the residual peak",
    )
    state_path: str | None = Field(
        default=None,
        description=("Optional deconvolution state NPZ path for resuming with additional cycles"),
    )
    method: Literal["sum", "mean"] = Field(
        default="sum",
        description="Integration method for the 2D preview images",
    )


class ImagePreviewPayload(BaseModel):
    """Normalized 2D preview and statistics for a cube product."""

    image: list[list[int]]
    stats: dict[str, Any]
    method: str


class DeconvolutionResponse(BaseModel):
    """Frontend response payload for iterative deconvolution previews."""

    dirty: ImagePreviewPayload
    component_model: ImagePreviewPayload
    restored: ImagePreviewPayload
    residual: ImagePreviewPayload
    reference_clean: ImagePreviewPayload | None = None
    convolved_reference: ImagePreviewPayload | None = None
    metadata: dict[str, Any]
